### PR TITLE
Fix eps in `Contrastive.backward`

### PR DESCRIPTION
**Pull request by asi1024**
Related to #7704.
In `Contrastive.backward`, eps(=`1e-7`) is added to `dist` to avoid zero-division. However, when the dtype of input arrays are `float16` it cause overflow because float16 arrays can't express `1e7`.

### CODE REVIEW

File: chainer/functions/loss/contrastive.py [hunk 1]
```diff
@@ -72,8 +72,9 @@ def backward(self, indexes, grad_outputs):
         dist = chainer.functions.repeat(dist[:, None], x_dim, axis=1)
         # avoid division by zero, 1e-7 is sufficiently small value that can be
         # represented even in half precision
+        eps = 5e-3 if dist.dtype == 'float16' else 1e-7
```
Comment (emcastillo) [inline chainer/functions/loss/contrastive.py]: I might be missing something but the comment says that 1e-7 can be represented in fp16.
I think that according to your PR description, the  comment should be updated too :)

Thanks

Comment (asi1024) [reply chainer/functions/loss/contrastive.py]: Updated comment.

**Comment by emcastillo** [comment]
Jenkins, test this please.

**Comment by emcastillo** [comment]
CI errors unrelated

**Comment by asi1024** [comment]
@emcastillo Could you backport this PR?

**Pull request merged**